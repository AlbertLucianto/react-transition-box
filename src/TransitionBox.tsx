import React, {
  useMemo,
  useRef,
  useState,
} from 'react';
import {
  createContainerStyle,
  innerStyle,
  objectStyle,
} from './styles';
import useResizeListener from './useResizeListener';

interface ITransitionBoxProps {
  children: JSX.Element[]|JSX.Element|string;
  timingFunction?: string;
  duration: number;
  [props: string]: any;
}

/**
 * 
 * @param param0 
 */
function TransitionBox({
  children,
  duration,
  timingFunction,
  style: propStyle,
  ...restProps
}: ITransitionBoxProps) {
  const ref = useRef<HTMLObjectElement>(null);
  const [style, setStyle] = useState({});
  const containerStyle = useMemo(
    () => createContainerStyle(duration, timingFunction),
    [duration, timingFunction],
  );

  const mergedStyle = useMemo(() => ({
    ...style,
    ...containerStyle,
    ...(propStyle || {}),
  }), [style, propStyle]);

  useResizeListener(ref, setStyle);

  return (
    <div style={mergedStyle} {...restProps}>
      <div style={innerStyle}>
        <object
          ref={ref}
          data="about:blank"
          aria-hidden={true}
          aria-label="resize-listener"
          style={objectStyle}
          tabIndex={-1}
          type="text/html"
        />
        {children}
      </div>
    </div>
  );
}

export default TransitionBox;
