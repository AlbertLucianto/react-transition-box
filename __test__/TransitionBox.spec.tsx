import React, { useMemo } from 'react';
import { create } from 'react-test-renderer';
import TransitionBox from '../src';

const waitFor = async (ms: number) => new Promise((resolve) => {
  setTimeout(resolve, ms);
});

interface IFooProps {
  width: number;
  height: number;
}

function Foo({ width, height }: IFooProps) {
  const style = useMemo(() => ({ width, height }), [width, height]);

  return (
    <TransitionBox
      duration={100}
      timingFunction="ease"
      className="foo__container"
    >
      <div style={style} className="abc" />
    </TransitionBox>
  );
}

describe('TransitionBox', () => {
  test('should render correctly', async (done) => {
    const style = {
      height: 200,
      width: 100,
    };

    const addListenerMock = jest.fn();
    const removeListenerMock = jest.fn();

    const el = create(
      <Foo width={style.width} height={style.height} />,
      {
        createNodeMock(element: JSX.Element) {
          if (element.type === 'object') {
            return {
              contentDocument: {
                defaultView: {
                  addEventListener(type: string, fn: () => void) {
                    addListenerMock(type);
                  },
                  removeEventListener(type: string, fn: () => void) {
                    removeListenerMock(type);
                  },
                },
              },
              offsetHeight: style.height,
              offsetWidth: style.width,
            };
          }

          return null;
        },
      },
    );
    expect(el.toJSON()).toMatchSnapshot();
    expect(addListenerMock).not.toHaveBeenCalled();
    expect(removeListenerMock).not.toHaveBeenCalled();

    expect(removeListenerMock).not.toHaveBeenCalled();

    style.width = 150;
    style.height = 100;
    el.update(<Foo width={style.width} height={style.height} />);
    expect(el.toJSON()).toMatchSnapshot();

    await waitFor(120);
    expect(el.toJSON()).toMatchSnapshot();

    el.unmount();
    expect(removeListenerMock).toHaveBeenNthCalledWith(1, 'resize');

    done();
  });
});
