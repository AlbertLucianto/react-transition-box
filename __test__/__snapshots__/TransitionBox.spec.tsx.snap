// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransitionBox should render correctly 1`] = `
<div
  className="foo__container"
  style={
    Object {
      "OTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "WebkitTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
    }
  }
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 200,
          "width": 100,
        }
      }
    />
  </div>
</div>
`;

exports[`TransitionBox should render correctly 2`] = `
<div
  className="foo__container"
  style={
    Object {
      "OTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "WebkitTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "height": 200,
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
      "width": 100,
    }
  }
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;

exports[`TransitionBox should render correctly 3`] = `
<div
  className="foo__container"
  style={
    Object {
      "OTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "WebkitTransition": "
    width 100ms ease,
    height 100ms ease
  ",
      "height": 200,
      "transition": "
    width 100ms ease,
    height 100ms ease
  ",
      "width": 100,
    }
  }
>
  <div
    style={
      Object {
        "height": "fit-content",
        "position": "relative",
        "width": "fit-content",
      }
    }
  >
    <object
      aria-hidden={true}
      aria-label="resize-listener"
      data="about:blank"
      style={
        Object {
          "display": "block",
          "height": "100%",
          "left": 0,
          "opacity": 0,
          "overflow": "hidden",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "width": "100%",
          "zIndex": -1,
        }
      }
      tabIndex={-1}
      type="text/html"
    />
    <div
      className="abc"
      style={
        Object {
          "height": 100,
          "width": 150,
        }
      }
    />
  </div>
</div>
`;
